src/controller/domain_handle.rs

```diff
@@ -69,10 +70,10 @@ impl<'a> BatchSendHandle<'a> {
                     };
                     dataflow::shard_by(key, self.dih.txs.len())
                 };
-                shard_writes[shard].push(r);
+                shard_writes.entry(shard).or_insert_with(Vec::new).push(r);
             }
 
-            for (s, rs) in shard_writes.drain(..).enumerate() {
```

Comment:
Instead of using a HashMap, you could just replace this line with:

```
for (s, rs) in shard_writes.drain(..).filter(|rs| !rs.is_empty()).enumerate() {	
```

Reply:
I think that would make the indices wrong? Since you're filtering out before you call `.enumerate()`. And filtering afterwards probably wouldn't work too well in terms of the borrow checker right (since you'd need to move out `rs`)?